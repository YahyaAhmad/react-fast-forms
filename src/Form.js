import React, { useState, useCallback, useEffect } from "react";
import PropTypes, { any, object, element, node } from "prop-types";
import { forEach, pickBy, omit } from "lodash";
import { validate, isNotValid } from "./utilities";
export const FormContext = React.createContext(null);

const Form = ({
  onSubmit = () => null,
  className,
  renderField,
  defaultValues = {},
  renderErrorMessage = message => (
    <div className="form-ErrorMessage">{message}</div>
  ),
  requiredErrorMessage = "This field is required",
  renderAllMessages = false,
  removeErrorsOnChange = false,
  omitInvalidValues = false,
  children,
  debug = false
}) => {
  const [data, setData] = useState({});
  const [errors, setErrors] = useState({});
  const [validators, setValidators] = useState({});
  const [fields, setFields] = useState([]);

  const setError = useCallback(
    (name, errorMessage) => {
      let newErrors = {};
      newErrors[name] = errorMessage;
      setErrors(prev => ({ ...prev, ...newErrors }));
    },
    [errors]
  );

  const handleChange = useCallback(
    (name, value) => {
      console.log(value);
      if (value == null && defaultValues[name]) {
        return;
      }
      if (removeErrorsOnChange) {
        setErrors({});
      }

      let newData = {};
      newData[name] = value;
      setData(prev => ({ ...prev, ...newData }));
    },
    [data]
  );

  const handleErrorMessage = (typeOfValidator, errorMessages) => {
    const errorMessage = errorMessages[typeOfValidator];
    switch (typeOfValidator) {
      case "required":
        return requiredErrorMessage;
      default:
        return errorMessage
          ? errorMessage
          : `Default error message of type ${typeOfValidator}`;
    }
  };

  const handleValidators = (fieldValidators, fieldName, errorMessages) => {
    let validated = true;
    forEach(fieldValidators, (rule, typeOfValidator) => {
      if (!validate(data[fieldName], rule, typeOfValidator)) {
        setError(fieldName, handleErrorMessage(typeOfValidator, errorMessages));
        validated = false;
        return false;
      }
    });
    return validated;
  };

  const handleAllValidators = useCallback(
    validators => {
      let validated = true;
      forEach(validators, validatorObject => {
        const fieldName = validatorObject.name;
        const fieldValidators = validatorObject.validators;
        const errorMessages = validatorObject.messages;
        const fieldValidated = handleValidators(
          fieldValidators,
          fieldName,
          errorMessages
        );
        if (!fieldValidated) {
          validated = false;
        }
        if (!renderAllMessages && !fieldValidated) {
          return false;
        }
      });

      return validated;
    },
    [data]
  );

  const handleSubmit = useCallback(
    e => {
      // Clear errors.
      setErrors({});

      // Prevent the from from refreshing.
      e.preventDefault();

      // Check and handle all validators.
      const validated = handleAllValidators(validators);
      // Pass the data to the onSubmit prop if there is no errors.
      if (validated) {
        let dataToSubmit = { ...data };
        if (omitInvalidValues) {
          // Remove any invalid value.
          dataToSubmit = pickBy(dataToSubmit, value => !isNotValid(value));
        }
        onSubmit(dataToSubmit);
      }
    },
    [errors, data, validators]
  );

  /**
   * Lets Field components register their validators
   *
   */
  const register = validatorObject => {
    let newValidators = {};
    newValidators[validatorObject.name] = validatorObject;
    setValidators(prev => ({ ...prev, ...newValidators }));
  };

  const handleDelete = (name, delta) => {
    let newData = { ...data };
    newData[name] = omit(newData[name], delta);
    setData(newData);
  };

  const onDelete = React.useRef(handleDelete);
  const onChange = React.useRef(handleChange);

  useEffect(() => {
    onDelete.current = handleDelete;
    onChange.current = handleChange;
  });

  const formContextValues = {
    onChange: onChange,
    onDelete: onDelete,
    data,
    errors,
    setError,
    removeErrorsOnChange,
    requiredErrorMessage,
    handleErrorMessage,
    renderAllMessages,
    clearErrors: () => setErrors({}),
    validators,
    handleValidators,
    register,
    renderErrorMessage
  };

  // Debug the changes on the data.
  useEffect(() => {
    if (debug) {
      console.log(data);
    }
  }, [data]);

  // Set the default values.
  useEffect(() => {
    setData(defaultValues);
  }, []);

  return (
    <FormContext.Provider value={formContextValues}>
      <form
        onSubmit={handleSubmit}
        className={["form-MainForm", className].join(" ")}
      >
        {children}
      </form>
    </FormContext.Provider>
  );
};

Form.propTypes = {
  onSubmit: PropTypes.func,
  className: PropTypes.string,
  children: PropTypes.node,
  debug: PropTypes.bool,
  renderField: PropTypes.func,
  renderErrorMessage: PropTypes.func,
  requiredErrorMessage: PropTypes.string,
  renderAllMessages: PropTypes.bool,
  validateOnChange: PropTypes.bool
};

export default Form;
