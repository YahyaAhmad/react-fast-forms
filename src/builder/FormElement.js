export default class FormElement {
  constructor(element) {
    this.element = element;
    this.props = {
      fields: {}
    };
    this.elementType = null;
    this.formProps = {};
  }
  getFormProps = () => {
    return this.formProps;
  };
  setFormProps = props => {
    this.formProps = props;
    return this;
  };
  getType = () => {
    return this.elementType;
  };

  setLabel = label => {
    this.props.label = label;
    return this;
  };

  getLabel = () => {
    return this.props.label;
  };

  setProp = (name, value) => {
    this.props[name] = value;
    return this;
  };

  getProps = () => {
    return this.props;
  };

  getComponent = () => {
    return this.element;
  };

  /**
   * Returns sub elements of the element.
   *
   * @returns {Element[]}
   */
  getElements = () => {
    return this.props.fields;
  };

  addField = (field, name) => {
    if (!this.props.fields) {
      this.props.fields = {};
    }
    this.props.fields[name] = field;
    return this;
  };
}
