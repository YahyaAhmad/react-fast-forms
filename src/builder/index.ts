import { isValidElement } from "react";
import GenericContainer from "../elements/GenericContainer";
import GenericField from "../elements/GenericField";

export interface ElementInterface<T> {
    label: string
    name: string
    component: React.FC
    type: string
    props: any
    getLabel: () => string
    setLabel: (label: string) => T
    getName: () => string
    setName: (name: string) => T
    getComponent: () => React.FC
    setComponent: (component: React.FC) => T
    getType: () => string
    setProps: (props: object) => T,
    getProps: () => object,
    setProp: (key: string, value: string) => T
}

export interface FieldInterface {
    validator: ValidatorFunction
    errorMessage: string
    value: any
    required: boolean
    rendered: boolean
    setValidator: (validator: ValidatorFunction) => Field
    validate: (value: any) => boolean
    setError: (error: string) => void
    getError: () => string
    setRequired: (required: boolean) => Field
    isRequired: () => boolean
    setRendered: (rendered: boolean) => Field
    isRendered: () => boolean
}

export interface ContainerInterface {
    fields: Array<Field>
    addField: (field: Field) => Container
    addFields: (fields: Array<Field>) => Container
    getFields: () => Array<Field>
}

type SetErrorFunction = (message: string) => void;
type ValidatorFunction = (value: string, setError: SetErrorFunction) => void;
export class Field implements FieldInterface, ElementInterface<Field> {
    required: boolean = false;
    rendered: boolean = true;
    label: string;
    name: string;
    component: React.FC
    type: string = "field"
    validator: ValidatorFunction = (value, setError) => null;
    errorMessage: string = null;
    value: any = null;
    props: any = {};
    getLabel = () => {
        return this.label;
    }
    setLabel = (label: string) => {
        this.label = label;
        return this;
    }
    setValidator = (validator: ValidatorFunction) => {
        this.validator = validator;
        return this;
    }
    getName = () => this.name;

    setName = (name: string) => {
        this.name = name;
        return this;
    }

    validate = (value: any) => {
        this.setError(null);
        this.validator(value, this.setError);
        if (this.getError()) {
            return false;
        } else {
            return true;
        }
    };

    setError = (error: string) => {
        this.errorMessage = error;
    }

    getError = () => this.errorMessage;

    getComponent = () => this.component

    setComponent = (component: React.FC) => {
        this.component = component;
        return this;
    }
    getType = () => this.type;

    setProps = (props: object) => {
        this.props = props;
        return this;
    }
    getProps = () => this.props;
    setProp = (key: string, value: string) => {
        this.props[key] = value;
        return this;
    }
    setRequired = (required: boolean = true) => {
        this.required = required;
        return this;
    };
    isRequired = () => this.required;
    setRendered = (rendered: boolean = true) => {
        this.rendered = rendered;
        return this;
    };
    isRendered = () => this.rendered;
}

export class Container implements ContainerInterface, ElementInterface<Container> {
    fields: Field[] = [];
    label: string;
    name: string;
    component: React.FC
    type: string = "container"
    props: any = {};
    getLabel = () => this.name;

    setLabel = (label: string) => {
        this.label = label;
        return this;
    };
    getName = () => this.name;
    setName = (name: string) => {
        this.name = name;
        return this;
    };
    getComponent = () => this.component

    setComponent = (component: React.FC) => {
        this.component = component;
        return this;
    }

    addField = (field: Field) => {
        this.fields.push(field);
        return this;
    };
    addFields = (fields: Field[]) => {
        this.fields.concat(fields);
        return this;
    };

    getFields = () => this.fields;
    getType = () => this.type;
    setProps = (props: object) => {
        this.props = props;
        return this;
    }
    getProps = () => this.props;
    setProp = (key: string, value: string) => {
        this.props[key] = value;
        return this;
    }
}

export function createContainer(component: React.FC = GenericContainer, name: string = null) {
    let container = new Container();
    container.setComponent(component);
    container.setName(name);
    return container;
}
type FormInputType = "text" | "password" | "select" | "date" | "textarea";

export function createField(component: React.FC | FormInputType, name: string) {
    if (!name) {
        throw new Error("There is a field without a name");
    }

    let field = new Field();
    if (typeof component === "string") {
        switch (component) {
            case "textarea":
                throw new Error("Textarea is not implemented");
                break;
            default:
                field.setProp("type", component);
        }
        component = GenericField;
    }

    if (typeof component === "function") {
        field.setComponent(component);
    } else {
        throw new Error("The component that was provided is not valid.");
    }

    field.setName(name);
    return field;
}