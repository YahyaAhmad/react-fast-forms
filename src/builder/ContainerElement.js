import { FormElement } from "./FormElement";
import { GenericContainer } from "..";

/**
 *
 * Creates a container
 *
 * @param {React.ComponentClass} container
 *
 * @returns {ContainerElement}
 */
export function createContainer(container) {
  let containerElement = container ? container : GenericContainer;
  let containerObject = new ContainerElement(containerElement);
  return containerObject;
}

export class ContainerElement extends FormElement {
  constructor(element) {
    super(element);
    this.elementType = "container";
  }
}
