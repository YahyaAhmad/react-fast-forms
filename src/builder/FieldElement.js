import FormElement from "./FormElement";

/**
 * Creates a new field element
 *
 * @param {Function} field
 * @param {string} name
 */
export function createField(field, name) {
  let fieldObject = new FieldElement(field, name);
  return fieldObject;
}

export default class FieldElement extends FormElement {
  constructor(element, name) {
    super(element);
    this.props.name = name;
    this.elementType = "field";
  }
  getName = () => {
    return this.props.name;
  };
  setEmptyValue(value) {
    this.props.emptyValue = value;
    return this;
  }
  getDefaultValue() {
    return this.props.defaultValue;
  }
  setDefaultValue(value) {
    this.props.defaultValue = value;
    return this;
  }
  setPlaceholder(placeholder) {
    this.props.options.placeholder = placeholder;
    return this;
  }

  setMultiple(multiple = true) {
    this.props.multiple = multiple;
    return this;
  }
  isMultiple() {
    return this.props.multiple;
  }
  setRequired(required = true) {
    this.props.required = required;
    return this;
  }
  isRequired() {
    return this.props.required;
  }
  setItems(items) {
    this.props.items = items;
    return this;
  }
  getItems() {
    return this.props.items;
  }

  setDependency(name) {
    this.dependsOn = name;
    return this;
  }
  setContainer(container) {
    this.container = container;
    return this;
  }
  getContainer() {
    return this.container ? this.container : EmptyContainer;
  }
  setContainerProps(props) {
    this.containerProps = props;
    return this;
  }
  getContainerProps() {
    return this.containerProps;
  }
  setContainerLabel(label) {
    this.containerLabel = label;
    return this;
  }
  getContainerLabel() {
    return this.containerLabel;
  }
  getDependency() {
    return this.dependsOn;
  }

  replaceDependencies(value) {
    Object.keys(this.props).map(key => {
      if (typeof this.props[key] === "object" && this.props[key] !== null) {
        this.props[key] = this.replaceDependencies(this.props[key]);
      } else if (typeof this.props[key] === "string") {
        this.props[key] = this.props[key].replace(
          `{${this.getDependency()}}`,
          value
        );
      }
    });
    return this;
  }
}
