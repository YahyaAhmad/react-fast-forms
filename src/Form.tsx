import { ContainerInterface, Field, Container } from './builder';
import FieldManager from './elements/FieldManager';
import * as React from 'react';
import PropTypes, { element } from "prop-types";
import ContainerManager from './elements/ContainerManager';
import { omit } from 'lodash';

interface FastFormProps {
    container: ContainerInterface
    onSubmit?: (data: object) => void
    debug?: boolean,
    onChange?: (name: string, value: any) => void
    attributes?: React.HTMLAttributes<Element>
}

export type Data = Record<string, any>

export const FormContext = React.createContext(null);

let addedFields: Array<Field> = [];

const Form = (props: FastFormProps) => {
    const { container, debug, onSubmit, onChange, attributes } = props;
    const [data, setData] = React.useState<Data>({});
    const [errors, setErrors] = React.useState<Data>({});
    const [render, setRender] = React.useState(null);
    const elements = container.getFields();
    const handleChange = (name: string, value: any) => {
        let newData = { ...data };
        newData[name] = value;
        setData(newData);
    }

    const handleSubmit = (e: React.MouseEvent) => {
        e.preventDefault();
        let submittedData = {};
        if (submittedData = validateForm()) {
            onSubmit(submittedData);

        } else {
            throw new Error("There are errors!");
        }
    }

    const isValid = (value: any) => !(value === null || value === undefined || value === "" || value === {} || value === "{}" || value === [])

    const setError = (name: string, errorMessage: string) => {
        let newErrors = { ...errors };
        newErrors[name] = errorMessage;
        setErrors(newErrors);
    }

    const validateForm = () => {
        let valid = true;
        setErrors({});
        let submittedData = { ...data };
        addedFields.map(field => {
            const name = field.getName();
            const value = submittedData[name];
            // Check if it is required
            if (!isValid(value)) {
                if (field.isRequired()) {
                    setError(name, "This field is required");
                    valid = false;
                }
                submittedData = omit(submittedData, name);
            }


            // Check the custom validator
            if (isValid(value)) {
                if (!field.validate(value)) {
                    setError(name, field.getError());
                }

            }
        });
        return valid ? submittedData : false;
    }

    const recursiveRender = (elements: (Field | Container)[]) => {
        let renderElements: React.ReactNode;
        elements.forEach((element) => {
            if (element instanceof Container) {
                const ContainerComponent = element.getComponent();
                let children = recursiveRender(element.getFields());
                renderElements = (
                    <>
                        {renderElements}
                        <ContainerManager render={children} containerProps={element.getProps()} label={element.getLabel()}>
                            <ContainerComponent />
                        </ContainerManager>
                    </>
                )
            } else if (element instanceof Field) {
                addedFields.push(element);
                renderElements = (
                    <>
                        {renderElements}
                        <FieldManager error={errors[element.getName()]} field={element} />
                    </>
                )
            }
        });
        return renderElements;
    }

    React.useEffect(() => {
        addedFields = [];

        setRender(recursiveRender(elements));

    }, [elements, errors]);

    const formProviderValues = {
        data: data,
        onChange: handleChange,
    }

    if (debug) {
        console.log(data);
    }


    return (
        <form {...attributes}>
            <FormContext.Provider value={formProviderValues}>
                {render}
            </FormContext.Provider>
            <button onClick={handleSubmit}>Submit</button>
        </form>
    );
}

Form.defaultProps = {
    onSubmit: (data: Record<string, any>) => {
        console.log("Successfully Submitted the values:");
        console.log(data);
    },
    onChange: (name: string, value: any) => { },
    debug: false,
    container: null,
}

Form.propTypes = {
    container: PropTypes.instanceOf(Container).isRequired
}

export default Form


