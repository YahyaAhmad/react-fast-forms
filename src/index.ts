import Form from './Form';
import { createContainer, createField } from './builder'
import { useField, useContainer } from './hooks';

export {
    Form,
    createContainer,
    createField,
    useField,
    useContainer
}