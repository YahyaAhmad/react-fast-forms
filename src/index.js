import { createField } from "./builder/FieldElement";
import { createContainer } from "./builder/ContainerElement";
import GenericContainer from "./elements/containers/GenericContainer";
import GenericField from "./elements/fields/GenericField";
import React, { Component } from "react";
import OptionsField from "./elements/fields/OptionsField";
import Form from "./forms/Form";

import CompositeField from "./elements/fields/CompositeField";

export {
  Form,
  createField,
  createContainer,
  CompositeField,
  GenericContainer,
  GenericField,
  OptionsField
};
