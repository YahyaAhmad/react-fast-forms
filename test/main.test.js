import { createField } from "../src/builder/FieldElement";
import { createContainer } from "../src/builder/ContainerElement";
import TestRenderer from "react-test-renderer";
import { Form } from "../src";
import { GenericContainer } from "../src";
import { GenericField } from "../src";
import React from "react";

describe("Field Element", () => {
  it("Field validator works as an object", () => {
    let field = createField(null, "someName");
    field.setValidator((value, setError) => {
      if (value == "error") {
        setError("There is an error");
      }
    });
    // Check with a correct value.
    field.validate("not error");
    expect(field.hasError()).toEqual(false);
    // Check with an incorrect value.
    field.validate("error");
    expect(field.hasError()).toEqual(true);
    // Check if the message is right.
    expect(field.getErrorMessage()).toEqual("There is an error");
  });

  it("Field dependency works", () => {
    let dependentField = createField(null, "dependent");
    dependentField.setDependency("dependency");
    dependentField.setLabel("My {dependency}");
    let newProps = dependentField.replaceDependencies("Field");
    expect(newProps.label).toEqual("My Field");
  });
});

describe("Form", () => {
  let mainContainer = createContainer(GenericContainer).setLabel(
    "My Container"
  );
  let field1 = createField(GenericField, "field1").setProp(
    "className",
    "field1"
  );
  let field2 = createField(GenericField, "field2")
    .setLabel("Test")
    .setRequired();
  mainContainer.addFields([field1, field2]);

  it("should render expectedly", () => {
    const form = TestRenderer.create(<Form fields={mainContainer} />).toJSON();
    expect(form).toMatchSnapshot();
  });

  it("Render the main container with 2 fields", () => {
    const form = TestRenderer.create(<Form fields={mainContainer} />);
    expect(form.toJSON().children[1].children.length).toEqual(2);
  });
  it("Should not submit when there is a field that is required but empty", () => {
    const form = TestRenderer.create(
      <Form onSubmit={values => (submitted = true)} fields={mainContainer} />
    );
    let submitted = false;
    const instance = form.getInstance();
    instance.handleSubmit();
    expect(submitted).toEqual(false);
  });
  it("Should submit if every required field is satisfied", () => {
    const form = TestRenderer.create(
      <Form onSubmit={values => (submitted = true)} fields={mainContainer} />
    );
    let submitted = false;
    const instance = form.getInstance();
    instance.handleChange("some value", "field2");
    instance.handleSubmit();
    expect(submitted).toEqual(true);
  });

  it("Field validator works in a form", () => {
    let submitted = false;

    let mainContainer = createContainer(GenericContainer).setLabel(
      "My Container"
    );
    let field1 = createField(GenericField, "normalField");
    let field2 = createField(GenericField, "customRequiredField").setValidator(
      (value, setError) => {
        if (value == "error value") {
          setError("There is an error");
        }
      }
    );
    mainContainer.addFields([field1, field2]);

    const form = TestRenderer.create(
      <Form onSubmit={value => (submitted = true)} fields={mainContainer} />
    );
    const instance = form.getInstance();
    instance.handleChange("some value", "normalField");
    instance.handleSubmit();
    expect(submitted).toEqual(true);
    instance.handleChange("error value", "customRequiredField");
    submitted = false;
    instance.handleSubmit();
    expect(submitted).toEqual(false);
    instance.handleChange("success value", "customRequiredField");
    instance.handleSubmit();
    expect(submitted).toEqual(true);
  });

  it("Data should change as expected", () => {
    const form = TestRenderer.create(
      <Form onSubmit={values => (submitted = true)} fields={mainContainer} />
    );
    const instance = form.getInstance();
    instance.handleChange("some value", "field1");
    expect(instance.state.data).toEqual({ field1: "some value" });
    instance.handleChange("other value", "field2");
    expect(instance.state.data).toEqual({
      field1: "some value",
      field2: "other value"
    });
  });

  it("Field A should not render when it is depndent on field B when its value is not defined", () => {
    let dependentField = createField(GenericField, "dependentField")
      .setLabel("Label is depndent on {field1}")
      .setProp("className", "dependent-field")
      .setDependency("field1");
    mainContainer.addField(dependentField);
    const form = TestRenderer.create(
      <Form onSubmit={values => (submitted = true)} fields={mainContainer} />
    );
    const root = form.root;
    let mainContainerRoot = form.root.findByProps({
      className: "form-container"
    });
    expect(
      mainContainerRoot.findAllByProps({ className: "dependent-field" })[0]
    ).toBeUndefined();
    form.getInstance().handleChange("some value", "field1");
    expect(
      mainContainerRoot.findAllByProps({ className: "dependent-field" })[0]
    ).toBeDefined();
  });

  it("Should omit empty or invalid values on submit", () => {
    let submittedData = {};
    let container = createContainer(GenericContainer).addField(
      createField(GenericField, "field1")
    );
    const form = TestRenderer.create(
      <Form onSubmit={values => (submittedData = values)} fields={container} />
    );
    const instance = form.getInstance();
    instance.handleChange("", "field1");
    instance.handleChange({}, "field2");
    instance.handleChange([], "field3");
    instance.handleChange(null, "field4");
    instance.handleChange(false, "correctValue1");
    instance.handleChange("some value", "correctValue2");

    expect(instance.state.data).toEqual({
      field1: "",
      field2: {},
      field3: [],
      field4: null,
      correctValue1: false,
      correctValue2: "some value"
    });
    instance.handleSubmit();
    expect(submittedData).toEqual({
      correctValue1: false,
      correctValue2: "some value"
    });
    instance.handleChange;
  });
});
