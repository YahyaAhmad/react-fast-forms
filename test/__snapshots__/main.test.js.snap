// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render expectedly 1`] = `
<div
  className="react-main-form"
>
  <label
    className="form-container-title"
  >
    My Container
  </label>
  <div
    className="form-container"
  >
    <div
      className="form-field-container field-field1-container"
    >
      <label
        className="form-field-label"
      />
      <input
        className="field1"
        onChange={[Function]}
        type="text"
      />
    </div>
    <div
      className="form-field-container field-field2-container"
    >
      <label
        className="form-field-label"
      >
        Test
      </label>
      <input
        onChange={[Function]}
        type="text"
      />
    </div>
  </div>
  <div
    className="form-actions"
  >
    <div
      className="form-action"
    >
      <button
        onClick={[Function]}
      >
        Submit
      </button>
    </div>
  </div>
</div>
`;
